### PR TITLE
first draft of schema management

### PR DESCRIPTION
**Pull request by mehranshakeri**
related to issue #26

### CODE REVIEW

File: core/schema-management.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Schema Management
+
+## Problem Description
+
+To enable data exchange between business parties involved in a Baseline Workflow (or only Workstep?) (in short, parties), there must be present a set of pre-agreed schemas.
```
Comment (humbitious) [inline core/schema-management.md:5]: Parties is capitalized. Don't need to say, "in short, parties".

Comment (humbitious) [reply core/schema-management.md:5]: Is it a set of pre-agreed schemas or a pre-agreed schema [pertaining to that single workstep/workflow...or "schema or schemas".  We should clarify what would constitute the need for multiple schema. Also would help to indicate what would constitute a typical schema that would need to be prearranged.  Separately, we should address what methods are best used to arrange the common schema.  What do translators like Mulesoft bring to this?  Which tools are available to make this easy? Can/should there be schema publishing via counterparty (DIDs?) to make this more standard and automatically discoverable?

File: core/schema-management.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Schema Management
+
+## Problem Description
+
+To enable data exchange between business parties involved in a Baseline Workflow (or only Workstep?) (in short, parties), there must be present a set of pre-agreed schemas. 
+
+Furthermore, at least identifying different versions of the same schema must be possible for each Baseline instance. A Baseline instance may support different versions of the same schema.
```
Comment (humbitious) [inline core/schema-management.md:7]: Might be good to elaborate on what a "baseline instance" is in this context.

File: core/schema-management.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## Schema
+  
+Minimum data model describing the structure of exchanged messages between different components and APIs. It must include list of mandatory fields/properties and their types (and their description?).
```
Comment (humbitious) [inline core/schema-management.md:11]: I don't think description is a "must".  But I could be wrong.  Suggest "must include list of mandatory fields/properties and their types. Should include description."

File: core/schema-management.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+## Agreement on Schema
+(in scope?)
+An agreement could be _explicit_ with stating the list of agreed upon versioned schemas (e.g. structured as a Merkle Tree with schemas as leaves) by all the parties or _implicit_ by their implementation.
```
Comment (humbitious) [inline core/schema-management.md:15]: This is the basis for an interesting discussion worth engaging in.

File: core/schema-management.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+(question: An explicit agreement could be seen as a commitment which makes Baseline processes (at least the intial setup) more expensive where an implicit one could be seen as a set of supported APIs which is already live but poses runtime incompatibility.)
+
+## Reaching an Agreement on Schema
+(in scope?)
+
+## Managing Different Schemas
+
+All the schemas must be available to all the Baseline instances involved in a Workflow (or only Workstep?).
+
+Schemas may be available publicly to everyone.
```
Comment (humbitious) [inline core/schema-management.md:26]: suggest just "may be made available publicly".  Don't need "to everyone".

File: core/schema-management.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+A Schema must be uniquely identifiable by the Baseline instances.
```
Comment (humbitious) [inline core/schema-management.md:28]: I'm unclear on the context of this requirement.

File: core/schema-management.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+## Schema Management Component
+
+Responsible to hold a registry (mapping) of which schemas are used in which working groups. (or Workflow/Wrokstep?)
```
Comment (humbitious) [inline core/schema-management.md:32]: Does this suggest that there be a standard component included in a baseline stack that records and maintains current and past schema subscriptions?

File: core/schema-management.md (hunk 1 above)
Comment (mehranshakeri) [reply core/schema-management.md:5]: I was thinking we will have a set of schemas and in each Baseline workstep/flow we will use a subset of them. (off-chain communication, on-chain transactions, etc.)
And I share the other questions you've raised.

File: core/schema-management.md (hunk 2 above)
Comment (mehranshakeri) [reply core/schema-management.md:7]: Agree, should I refer to this? [2.6 Baseline Protocol Instance](https://github.com/ethereum-oasis/baseline-standard/pull/50/files?short_path=6db9664#diff-6db9664e3b55a0ca9841f1be1ca73aec567763fba50a0cc4dcaa4514ec88be4d)

I think I need to wait for other PR to be merged and use the proper link.

File: core/schema-management.md (hunk 4 above)
Comment (mehranshakeri) [reply core/schema-management.md:15]: My current position is that we need to foresee/expect both. But will it be in scope of standard or not is another question.

File: core/schema-management.md (hunk 6 above)
Comment (mehranshakeri) [reply core/schema-management.md:28]: My understanding is that a Baseline Protocol Instance will be reused for different workgroups/steps. Each group/step will use different schema(s). That's why we need a component to manage these schemas and also they must be uniquely identifiable to be addresses. Isn't this the main reason to have this brainstorming in the first place?

File: core/schema-management.md (hunk 7 above)
Comment (mehranshakeri) [reply core/schema-management.md:32]: At least currently in the brainstorming level, yes.

**Review by Therecanbeonlyone1969 — CHANGES_REQUESTED** [review]
Not only data but also functional.

`data exchange and functional interoperability`

**Review by Therecanbeonlyone1969 — CHANGES_REQUESTED** [review]
Should be 

`standardized and discoverable`

since discoverability is key. Also, we should consider saying `ontologically standardized` so not only standardized but standardized within a specific context such as supply chain, health, insurance ...

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
Which section in the standard does this fit in? ... might be good to have the right section number in the header.

**Comment by Therecanbeonlyone1969** [comment]
@mehranshakeri Would recommend using JSON-LD schemas and utilize context files for data schema management and subsequent validation through a JSON-LD checker.

Example data schema here: 
`{
    "$id": "https://w3id.org/traceability/schemas/LegalEntityIdentifierCredential.json",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$comment": "{\"term\": \"LegalEntityIdentifierCredential\", \"@id\": \"https://w3id.org/traceability#LegalEntityIdentifierCredential\"}",
    "title": "LegalEntityIdentifierCredential",
    "description": "GLEIF verifiable Legal Entity Identifier Credential (verifiable leiCode Credential)",
    "type": "object",
    "properties": {
        "@context": {
            "type": "array"
        },
        "type": {
            "oneOf": [
                {
                    "type": "string"
                },
                {
                    "type": "array"
                }
            ]
        },
        "leiCode": {
            "$comment": "{\"term\": \"leiCode\", \"@id\": \"https://schema.org/leiCode\"}",
            "title": "leiCode",
            "description": "leiCode - An organization identifier that uniquely identifies a legal entity as defined in ISO 17442",
            "type": "string"
        },
        "certificateName": {
            "$comment": "{\"term\": \"certificateName\", \"@id\": \"https://schema.org/name\"}",
            "title": "certificateName",
            "description": "Specific name of issued legal entity certificate",
            "type": "string"
        }
    },
    "required": [
        "leiCode",
        "certificateName"
    ],
    "additionalProperties": false,
    "examples": []
}`

context file example:
`
{
  "@context": {
    "@version": 1.1,
    "name": "http://schema.org/name",
    "description": "http://schema.org/description",
    "identifier": "http://schema.org/identifier",
    "image": {
      "@id": "http://schema.org/image",
      "@type": "@id"
    },
    "id": "@id",
    "type": "@type",
"LegalEntityIdentifierCredential": {
      "@id": "https://w3id.org/traceability#LegalEntityIdentifierCredential",
      "@context": {
        "leiCode": {
          "@id": "https://schema.org/leiCode"
        },
        "certificateName": {
          "@id": "https://schema.org/name"
        }
      }
    }
}
`
We also need application interop as this is just like data interop:

example of an app spec within a VC's `credentialSubject`:

`
{
  "@context": "https://baseline.org/baseline-business-services/spec/v1",
  "service": [{
    "id": "did:ethr:dc2bfb8d330a75bd6368468ad36dcd5a12ed4de69fd3b851d7dfc6f6efcd3912#keys-1",
    "type": "WholeSaleVoice_ACME",
    "serviceEndpoint": "https://u0yzz5mtif-u0ss36yzrk.us0-aws.acme.io/"
  }],
  "SupportedApplications": [{
  "BaselineApplicationType": "WholeSaleVoice",
  "BaselineApplicationName": "WholeSaleVoice_NA",
  "BaselineApplicationID": "did:ethr:ef2bfb8d330a75bd6368468ad36dcd5a12ed4de69fd3b851d7dfc6f6efcd4012",
  "BaselineApplicationProvider": "did:ethr:ef2bfb8d330a75bd6368468ad36dcd5a12ed4de69fd3b851d7dfc6f6efcd3912",
  "BaselineApplicationVersion": "1.0",
  "DLTSupported": ["Hyperledger_Besu_1.4.1", "Hyperledger_Fabric_2.0"}],
  "ApplicationAPIspecs": "https://u0yzz5mtif-u0ss36yzrk.us0-aws.acme.io/swaggerapis/?swagger?from=account1?source=wholesalevoicev1.0.json",
  "RPCApplicationEndpoint": "https://u0yzz5mtif-u0ss36yzrk-rpct.us0-aws.acmer.io",
  "WebhooksApplicationEndpoint": "https://u0yzz5mtif-u0ss36yzrk-connVCt.us0-aws.acmer.io"
  }],
  "created": "2020-10-10T17:00:00Z",
  "updated": "2021-10-17T02:41:00Z"
  }
`
The above needs to be discoverable -> use DID docs for that where the VC with the spec is hosted at a service endpoint from the DID doc e.g.:
`
"service": [{
    "id": "did:ethr:dc2bfb8d330a75bd6368468ad36dcd5a12ed4de69fd3b851d7dfc6f6efcd3912#keys-1",
    "type": "WholeSaleVoice_ACME",
    "serviceEndpoint": "https://u0yzz5mtif-u0ss36yzrk.us0-aws.acme.io/"
  }]
`
Then once you have that you need service negotiation since not everyone supports all versions and DLTs etc.

see example here taken from a telco context: 
Picture1

**Comment by Therecanbeonlyone1969** [comment]
@mehranshakeri given the nascent stage of the PR, I would recommend to covert this to a draft PR until it is fully PR ready. What do you think?

File: core/schema-management.md (hunk 1 above)
Comment (kthomas) [reply core/schema-management.md]: > I was thinking we will have a set of schemas and in each Baseline workstep/flow we will use a subset of them. (off-chain communication, on-chain transactions, etc.)
> And I share the other questions you've raised.

A lot of the schemas are actually well-known and as such, I don't think this necessarily belongs in the core specification. Organizations know how to marshal to and from well-known data models using whatever tools they currently use.

If you are using something that isn't a known format, a second option that is basically "generic/unchecked" could leverage some schema registry as you describe. For the most by-the-book cases though, a stack does not have to implement this if it knows what schema is supported for any given workstep...

Can call out how this works in the BRI-1 impl and API specification, but it is based on this screenshot, which does associate a model (or generic/unchecked one, which would require some further negotiation not necessarily required by the standard):

image

File: core/schema-management.md (hunk 6 above)
Comment (kthomas) [reply core/schema-management.md:28]: See above response @humbitious

File: core/schema-management.md (hunk 7 above)
Comment (kthomas) [reply core/schema-management.md:32]: See above response @humbitious @mehranshakeri

**Comment by kthomas** [comment]
Can we agree to discussing this requirement in the context of "unchecked" schemas as described in this comment?

https://github.com/ethereum-oasis/baseline-standard/pull/52#discussion_r623478415

**Comment by mehranshakeri** [comment]
Can everyone please comment on the problem description? I want to make sure we all agree on the problem, why it needs to be addressed as part of core specs and what happens if we don't?

e.g. if prior to any Baseline transactions, Parties need to sign [Commercial Agreements](https://github.com/ethereum-oasis/baseline-standard/blob/main/core/baseline-core-v1.0-psd01.md#24-commercially-and-legally-binding-documents), it means there will be an initial manual setup and during that phase they can agree on a schema set, configure their Baseline stacks if necessary and let it run. Do we really need an automated negotiation as part of core specs or are we over engineering? In this case, we only need to standardize minimum schemas for different APIs and objects which is happening in BRI.
@kthomas Are we talking about the same thing?
cc @Therecanbeonlyone1969 @humbitious

**Comment by Therecanbeonlyone1969** [comment]
@mehranshakeri ... negotiation can be as simple as storing acceptable configurations and schemas at `.well-known` under e.g. baslineconfiguration.json or .jsonld ... that would allow entities, for example, to start baselining by simply hitting a published endpoint to create an account etc. -> enables enterprise IT departments to easily build services and integrate them in their architecture without having to rely on 3rd parties.

As to commercial agreements, only if you are actually doing commerce, do you need a commercial agreement signed outside of the baseline context -- otherwise go ahead without one. The digital fingerprint of the commercial agreement aka hash value can be used as an initial input value into the baselined commercial contract.

**Comment by kthomas** [comment]
> @mehranshakeri ... negotiation can be as simple as storing acceptable configurations and schemas at `.well-known` under e.g. baslineconfiguration.json or .jsonld ... that would allow entities, for example, to start baselining by simply hitting a published endpoint to create an account etc. -> enables enterprise IT departments to easily build services and integrate them in their architecture without having to rely on 3rd parties.
> 
> As to commercial agreements, only if you are actually doing commerce, do you need a commercial agreement signed outside of the baseline context -- otherwise go ahead without one. The digital fingerprint of the commercial agreement aka hash value can be used as an initial input value into the baselined commercial contract.

+1 -- the agreements already exist today. Just memorialize them by reference and be done. No engineering needed...

As far as counterparty resolution and handshaking, we can describe the data models and APIs we have created to do this in PR #1.

cc: @LucasRodriguez

**Comment by mehranshakeri** [comment]
> @mehranshakeri ... negotiation can be as simple as storing acceptable configurations and schemas at `.well-known` under e.g. baslineconfiguration.json or .jsonld ... that would allow entities, for example, to start baselining by simply hitting a published endpoint to create an account etc. -> enables enterprise IT departments to easily build services and integrate them in their architecture without having to rely on 3rd parties.
> 
> As to commercial agreements, only if you are actually doing commerce, do you need a commercial agreement signed outside of the baseline context -- otherwise go ahead without one. The digital fingerprint of the commercial agreement aka hash value can be used as an initial input value into the baselined commercial contract.

Then can we say this is more of an implementation details that could be recommended but not a core spec?

File: core/schema-management.md (hunk 1 above)
Comment (mohanramanujam) [reply core/schema-management.md]: 1) is the negotiation and definition of a set of schemas that a workgroup agrees upon outside the scope of the spec? who and what decides informational needs by a dynamic set of participants or into a evolving workgroup?
2) [Un]Marshaling takes care of messaging transparency but would parties not still need a dictionary that represents the schemas they are working with? 

New to these discussions ... learning. Apologies if I am not quite understanding the context.

**Pull request closed without merging**